src/main/java/bio/terra/service/snapshot/SnapshotMapTableDao.java

```diff
@@ -62,73 +65,73 @@ protected void createColumns(UUID tableId, Collection<SnapshotMapColumn> columns
+          UUID mapColumnId = rs.getObject("map_column_id", UUID.class);
+          UUID fromColumnId = rs.getObject("from_column_id", UUID.class);
+          UUID toColumnId = rs.getObject("to_column_id", UUID.class);
+
+          DatasetTable datasetTable = (DatasetTable) getTable(fromTableId, source.getDataset());
+
+          SnapshotTable snapshotTable = (SnapshotTable) getTable(toTableId, snapshot);
+
+          SnapshotMapTable snapshotMapTable =
+              mappingMap.computeIfAbsent(
+                  mapTableId,
+                  id ->
+                      new SnapshotMapTable()
+                          .id(id)
+                          .fromTable(datasetTable)
+                          .toTable(snapshotTable)
+                          .snapshotMapColumns(new ArrayList<>()));
+
+          snapshotMapTable
+              .getSnapshotMapColumns()
+              .add(
+                  new SnapshotMapColumn()
+                      .id(mapColumnId)
+                      .fromColumn(getColumn(datasetTable, fromColumnId))
+                      .toColumn(getColumn(snapshotTable, toColumnId)));
+
+          return snapshotMapTable;
+        });
+
+    return List.copyOf(mappingMap.values());
   }
 
-  public List<SnapshotMapColumn> retrieveMapColumns(
-      UUID mapTableId, Table fromTable, Table toTable) {
-    String sql =
-        "SELECT id, from_column_id, to_column_id"
-            + " FROM snapshot_map_column WHERE map_table_id = :map_table_id";
-
-    List<SnapshotMapColumn> mapColumns =
-        jdbcTemplate.query(
-            sql,
-            new MapSqlParameterSource().addValue("map_table_id", mapTableId),
-            (rs, rowNum) -> {
-              UUID fromId = rs.getObject("from_column_id", UUID.class);
-              Optional<Column> datasetColumn = fromTable.getColumnById(fromId);
-              if (!datasetColumn.isPresent()) {
-                throw new CorruptMetadataException(
-                    "Dataset column referenced by snapshot source map column was not found");
-              }
-
-              UUID toId = rs.getObject("to_column_id", UUID.class);
-              Optional<Column> snapshotColumn = toTable.getColumnById(toId);
-              if (!snapshotColumn.isPresent()) {
-                throw new CorruptMetadataException(
-                    "Snapshot column referenced by snapshot source map column was not found");
-              }
-
-              return new SnapshotMapColumn()
-                  .id(rs.getObject("from_column_id", UUID.class))
-                  .fromColumn(datasetColumn.get())
-                  .toColumn(snapshotColumn.get());
-            });
-
-    return mapColumns;
+  private Table getTable(UUID tableId, FSContainerInterface container)
+      throws CorruptMetadataException {
+    var table =
+        switch (container.getCollectionType()) {
+          case DATASET -> ((Dataset) container).getTableById(tableId);
```

Comment:
Why is `getTableById()` not a method on `FSContainerInterface`? `Dataset` and `Snapshot` are the only subclasses, and they both have this method, so it seems like it wouldn't hurt to add it to the base type, and would let us clean up code like this.

Reply:
Ah, I tried to get it to work.  Unfortunately, they both return their own `Table` type objects and even with generic type annotations, because of type erasure, you can't set it at the interface level...it just knows that it returns an untyped `Optional`